src/main/java/org/truffleruby/core/encoding/EncodingNodes.java

```diff
@@ -741,51 +746,60 @@ protected int getIndex(RubyEncoding encoding) {
     }
 
     // MRI: rb_enc_check_str / rb_encoding_check
-    public abstract static class CheckRopeEncodingNode extends RubyBaseNode {
+    @Primitive(name = "encoding_ensure_compatible_str")
+    @ImportStatic(RubyGuards.class)
+    public abstract static class CheckRopeEncodingNode extends PrimitiveArrayArgumentsNode {
```

Comment:
I think it'd be best if the node is named like the Primitive, so it's clear it's only to expose it to Ruby code.

Reply:
Ah but actually it's used by both?
I think two different nodes are better then, to keep the clear `executeCheckEncoding(RopeWithEncoding first, RopeWithEncoding second)` for Java usages.

Reply:
BTW as a note for the interpreter it would likely be best to avoid using `RopeWithEncoding` for encoding checks altogether as that likely causes extra allocations in interpreter (it seems hard for the interpreter to prove `RopeWithEncoding` does not escape).

Reply:
Okay. I asked about this in the GraalVM Slack. You'd end up with a primitive node that just delegates all of its work to some other node. I find that more annoying than helpful. I think part of what made this code difficult to manage before was there were so many nodes related encoding negotiation and it wasn't clear what the entry point should be for any given caller.

Reply:
If the input are the same, I think having a `@Primitive` also used internally is completely fine (I think there is a case of that in one of the commits).
What's different here is they use different inputs, Object (Ruby String) vs RopeWithEncoding.
That feels bad because Java code might easily pass something else than a RopeWithEncoding or a String and then it'd just fail at runtime.
Also from a memory footprint point of view it's less optimal than separate nodes (e.g., extra RubyStringLibrary fields for Java usages).
I think having 2 nodes next to each other would be best here, and maybe a comment to clarify why we have 2 nodes (different inputs: Ruby Strings vs RopeWithEncoding)

Reply:
Fair point. I don't really like how some of this works out. The difference between a `RopeWithEncoding` and `RubyString` is fairly nuanced. I understand why we have it, but it feels like a hack we need because we have two different string classes and they're inconvenient to work with from Java.

I'll try splitting it into two nodes and see what it looks like. I'm not wild about having two different nodes that do the same thing but with different input types, given we have specializations to handle that, but working with `Object` is admittedly not ideal. I had played with adding a "helper" execute method and with `@CreateCast` to unify the types, but didn't want to incur the overhead on the `RopeWithEncoding` cases in the interpreter.

Anyway, I'll split into two and see how it looks now. I'm biased towards reducing the number of node classes related to encoding negotiation to reduce confusion, but maybe I went too far or maybe it's not too bad now.

Reply:
If we had a more precise type for Ruby Strings, we could have two execute methods like
`execute(RubyString, RubyString)` and `execute(RopeWithEncoding, RopeWithEncoding)` and then it would probably be clear enough, but I guess that doesn't work due to Ruby Strings having Object has the only common supertype (the second execute would include the Object specialization).

I think in genernal it's confusing if inputs are so different.
The only other place AFAIK where it's done is UndefMethodNode, but there both arguments are Ruby-level values, and there is a comment explaining it.

I think the several fields (RubyStringLibrary but also source-related fields) saved by making it a RubyBaseNode and not a PrimitiveArrayArgumentsNode for the RopeWithEncoding node is worth it.